Add GDF2 class tests

dcslagel opened the pull request:
#### Description 

This pull request adds a 2 GDF2 specific tests  to improve test coverage

- Add separate test file for GDF2 specific tests
- Add test for GDF2's repr which will include the number of records
- Add test for GDF2.get_column_definitions()

#### Tests

All tests pass

pytest results before this pull-request:
```
Name                    Stmts   Miss  Cover
-------------------------------------------
aseg_gdf2/__init__.py       2      0   100%
aseg_gdf2/gdf2.py         270     54    80%
-------------------------------------------
TOTAL                     272     54    80%
```

pytest results after this pull-request:

```
Name                    Stmts   Miss  Cover
-------------------------------------------
aseg_gdf2/__init__.py       2      0   100%
aseg_gdf2/gdf2.py         270     38    86%
-------------------------------------------
TOTAL                     272     38    86%
```

Let me know if this change could be accepted (or rejected) or
needs some additional changes before being approved and merged.

Thank you,
DC

dcslagel committed:
Add GDF2 class tests

- Add separate test file for GDF2 specific tests
- Add test for GDF2's repr which will include the number of records
- Add test for GDF2.get_column_definitions()

kinverarity1 reviewed (commented):
Hey @dcslagel  - thanks so much for looking at the test coverage!

One suggestion here though here - `0x10cfdead0` will change every time right, isn't it the memory location? We should match it in the text with a regex or just do a different test to avoid it.

dcslagel commented:
@kinverarity1, 

Good catch.  :-)  

It is weird and interesting, the test keeps passing but the memory location should have changed.  
Yep, I'll dig into it,  make the fix and post it in the next few days.

DC

dcslagel committed:
Fix GDF2 repr test

- Fix broken assert statements
- change __repr__ call to check self.nrecords instead of self._nrecords
  nrecords will call the nrecords() method and check the current number
  of records.

dcslagel commented:
@kinverarity1 , 

Commit 4af393d fixes the broken assert tests:
- It corrects the assert syntax in the test_gdf_class tests. (The incorrect syntax was causing the tests to always pass).
- Changes the test_repr_1() test to check only that the test ends with the correct number of records.

In addition, this change changes GDF2's __repr__() method to use self.nrecords rather than self._nrecords.  This is because self.nrecords() need to run at least once to report the correct number of nrecords.   

The Codacy/PR Quality Review is reporting the asserts as bad because they will be removed when the python code is compiled to object code, but this shouldn't be a problem for the test code and asserts are used in the other parts of the asset_gdf2 test suite.  If you agree, then it would be good to see if Codacy can ignore the tests directory for the assert check.  However, if the test_suite should change to use some other tool that assert for testing let me know, and I'll make the updates to this pull-request.

Let me know if this pull-request need further improvement to be accepted.

Thanks!,
DC

kinverarity1 commented on aseg_gdf2/gdf2.py:
```diff
@@ -83,7 +83,7 @@ def __init__(self, dfn_filename, method="whitespace", engine="pandas", **kwargs)
     def __repr__(self):
         r = super().__repr__()
-        nrecords = self._nrecords
+        nrecords = self.nrecords
```
Can we revert this? `self.nrecords` takes a long time for very large files, which is why it is behind a property method. It shouldn't be calculated until the user explicitly requests it or a method needs it.

dcslagel replied:
This change is made in commit a1c6d0b.  I also added some notes to make the gdf.nrecords delayed resolution purpose more explicit.  Let me know if the notes need to be improved or removed.

kinverarity1 commented on tests/test_gdf2_class.py:
```diff
@@ -0,0 +1,37 @@
+    GDF2's repr will display the number of records: nrecords
+    """
+    gdf = aseg_gdf2.read(data_src_1)
+    res = gdf.__repr__()
```
If you revert the change to `GDF2.__repr__`, then all we need here is to insert something like `blahblah = gdf.nrecords` between lines 23 and 24, which will trigger the calculation of the number of rows.

In fact you could probably test the entire behaviour here: initially `repr(gdf)` should report nrecords=? and only after the access of the property would it report the number of rows as the current test checks.

dcslagel replied:
This change is made in commit a1c6d0.  The steps and asserts in test_repr_1() are arranged as you suggested so that the whole behavior is demonstrated.  Let me know if it needs additional improvement.

kinverarity1 reviewed (changes requested):
Thanks @dcslagel see comments.

dcslagel committed:
Improve GDF repr test per review

- Revert call to nrecord to  _nrecord so that nrecord is only called
  when needed because it will take time on larger files.
- Add code and asserts to demonstrate the _nrecord/nrecord behavior in
  gdf.__repr__() calls

dcslagel commented:
Hi @kinverarity1 ,

I'm just following up on old pull-requests.  I think that I've made all the changes requested.  Let me know if there are further concerns to be addressed.  (It looks like the Codacy/PR Quality Review failed but it is not displaying the error for me.  Let me know if there is something I can fix there.)

Thank you,
DC

kinverarity1 reviewed (approved):
Thank you! So sorry that it has taken me so long to get to this!
